### DISCUSSION TITLE
how to create an appropriate temp vector?

### DISCUSSION THREAD

**Discussion by bricerebsamen** [NVIDIA/cccl#1033] [Thrust]
hello

I have a function that uses thrust and does something like this:

```
void foo(Point* points, int num_pts) {
  thrust::device_vector<int> tags(num_pts); // temp data
  thrust::fill(tags.begin(), tags.end(), 0);
  // .. work on tags and points together using a zip iterator and thrust::sort
}
```

As it is, it's designed to work on the GPU only, because it uses a device_vector. So it assumes `points` is a device pointer. How could I  modify this function so that it could accept either a device or host vector for points and work on cpu or gpu accordingly.

perhaps the new signature could take iterators instead, which would encode "where" the input points are located
```
void foo(PointIterator points_bgn, PointIterator points_end);
```

related to https://stackoverflow.com/questions/76597522/thrust-execution-policy-with-device-vector

**Comment by gevtushenko** [comment]
Hello @bricerebsamen! 

You could try template template parameters to achieve that:

```cpp
template <template <class, class> class VectorT, 
          template <class> class AllocatorT,
          class T> 
void foo(VectorT<T, AllocatorT<T>> &vec) {
  using vec_t = VectorT<T, AllocatorT<T>>;

  vec_t tmp(vec.size() /* size */, 0 /* fill value */);
}

int main() {
  {
    thrust::host_vector<double> points(100);
    foo(points);
  }

  {
    thrust::device_vector<double> points(100);
    foo(points);
  }
}
```

**Reply by miscco** [reply]
So in most cases it would be a bit of an overkill to go the full template template way. If you do not need the access to all the template arguments then just making it a simple template suffices:
```cpp
template <class Container> 
void foo(Container& vec) {
  Container tmp(vec.size() /* size */, 0 /* fill value */);
}
```

**Comment by pauleonix** [comment]
Just FYI, the `vector` (both Thrust and STL) constructor taking only the number of elements always value-initializes these elements, i.e. `tags` doesn't need to be filled with `0` after construction. If you want to be more explicit you can use the constructor with an explicit value, i.e. `thrust::device_vector<int> tags(num_pts, 0);`. Filling it afterwards is a waste (and at least for `device_vector` in wont be optimized away either).
